src/System.Text.Primitives/System/Text/FormattingData.cs

```diff
@@ -12,19 +12,309 @@ public struct FormattingData
+                {
+                    skippedValues++;
+                    continue;
+                }
+                byte temp = _digitsAndSymbols[i][0];
+                int j;
+                // the insertion sort loop. This is pretty standard
+                for (j = i - skippedValues - 1; j >= 0 && _parsingTrieList[j].ByteValue > temp; j--)
+                {
+                    if (j + 1 == _parsingTrieList.Count) // if we're at the end of the list, we append it
+                        _parsingTrieList.Add(_parsingTrieList[j]);
+                    else // otherwise, we swap
+                        _parsingTrieList[j + 1] = _parsingTrieList[j];
+                }
+                if (_parsingTrieList.Count < 1) // if this is the first element, we should just add it to the list
+                {
+                    _parsingTrieList.Add(new TrieNode(temp, i));
+                }
+                else if (j == -1) // we have to check for this before the collision check so that we don't throw an index error
+                {
+                    _parsingTrieList[j + 1] = new TrieNode(temp, i); // this completes the swap
+                }
+                else if (_parsingTrieList[j].ByteValue == temp) // if a collision occurred
+                {
+                    CreateChildNode(_parsingTrieList[j], i, 1); // we need to create a child node (start building the trie)
+                    skippedValues++;
+                }
+                else if (j >= _parsingTrieList.Count - 1) // if we're at the end of the list, then we just append it
+                {
+                    _parsingTrieList.Add(new TrieNode(temp, i));
+                }
+                else
+                {
+                    _parsingTrieList[j + 1] = new TrieNode(temp, i); // this completes the swap
+                }
+            }
+        }
+        private void CreateChildNode(TrieNode node, int newIndex, int level)
+        {
+            CreateChildNode(node, newIndex, node.ChildIndices[0], level);
+        }
+        private void CreateChildNode(TrieNode node, int newIndex, int oldIndex, int level)
+        {
+            byte? newByteValue = DisambiguateByteValue(newIndex, level); // this is the value we're attempting to place in the trie
+
+            // If the node is a leaf, we know we have reached the bottom of the trie.
+            if (node.IsLeaf)
+            {
+                node.IsLeaf = false; // this is now a branch, not a leaf
+                byte? oldByteValue = DisambiguateByteValue(oldIndex, level);
+                if (newByteValue == oldByteValue)   // this means there's another collision, so we have to go down one level more
+                {
+                    if (oldByteValue == null)   // if both values are null, we have identical code units, which is bad
+                    {
+                        throw new ArgumentException("Invalid mapping: two values map to the same code unit.");
```

Comment:
I don't think a developer callign this API will be able to figure out from the message what's going on. Could the message say something about two digits and/or symbols being encoded with the same sequences of bytes?